src/RoadRegistry.Syndication.ProjectionHost/AtomFeedProcessor.cs

```diff
@@ -0,0 +1,242 @@
+                                                context.ChangeTracker.DetectChanges();
+                                                await context.SaveChangesAsync(_messagePumpCancellation.Token).ConfigureAwait(false);
+                                                await context.DisposeAsync().ConfigureAwait(false);
+
+                                                context = dbContextFactory();
+                                                context.ChangeTracker.AutoDetectChangesEnabled = false;
+                                                observedMessageCount = 0;
+                                            }
+                                        }
+
+                                        entries = (await reader.ReadEntriesAsync(
+                                                feedConfiguration.Uri,
+                                                catchUpPosition,
+                                                feedConfiguration.UserName,
+                                                feedConfiguration.Password,
+                                                true,
+                                                false))
+                                            .ToList();
+                                    }
+
+                                    if (observedMessageCount > 0) // case where we just read the last page and pending work in memory needs to be flushed
+                                    {
+                                        logger.LogInformation(
+                                            "[{Context}] Flushing catch up position of {Position} and persisting changes ...",
+                                            typeof(TSyndicationContent).Name, catchUpPosition);
+                                        await context
+                                            .UpdateProjectionState(
+                                                roadRegistrySyndicationProjectionHost,
+                                                catchUpPosition,
+                                                _messagePumpCancellation.Token)
+                                            .ConfigureAwait(false);
+                                        context.ChangeTracker.DetectChanges();
+                                        await context.SaveChangesAsync(_messagePumpCancellation.Token).ConfigureAwait(false);
+                                    }
+
+                                    await context.DisposeAsync().ConfigureAwait(false);
+
+                                    await scheduler.Schedule(async token =>
+                                    {
+                                        if (!_messagePumpCancellation.IsCancellationRequested)
+                                        {
+                                            await using (var resumeContext = dbContextFactory())
+                                            {
+                                                var projection =
```

Comment:
Bit strange you're doing lookups in the callback - writing to a message channel has almost no reason to fail, accessing a database has many reasons to fail (network, database). You've also been keeping track of the `catchUpPosition` so you don't need a lookup here.

Reply:
Reason was the `catchupPosition` var could theoretically be modified outside of the scope. In practice it won't happen though, so you're right.